File: flashInfo/flash_info.py
#----made by zsir----
import sys

argc = len(sys.argv)
if argc < 2:
    print "read NANDIFO.nni and write nandinfo.txt"
    input_file = open('./NANDINFO.nni','rb')
    output_file = open('./nandinfo.txt','w')
elif argc == 2:
    print "read %s and write nandinfo.txt" % sys.argv[1]
    input_file = open(sys.argv[1],'rb')
    output_file = open('./nandinfo.txt','w')
elif argc == 3:
    print "read %s and write %s" % (sys.argv[1], sys.argv[2])
    input_file = open(sys.argv[1],'rb')
    output_file = open(sys.argv[2],'w')
else:
    print "usage: python flash_info.py"
    print "or + input_file"
    print "or + input_file + output_file"
    sys.exit()

#print sys.argv[1]

BLOCK_SIZE = 512
num = 0
while True:
    content = input_file.read(BLOCK_SIZE)
    if not content:
        break
    else:
        nand_id_number = int(content[0x10].encode('hex'), 16)
        for ix in range(nand_id_number):
            if ix == 0:
                output_file.write("nand id:")
                output_file.write(content[0x11 + ix].encode('hex'))
            elif ix == nand_id_number - 1:
                output_file.write("-")
                output_file.write(content[0x11 + ix].encode('hex'))
                output_file.write("\r\n")
            else:
                output_file.write("-")
                output_file.write(content[0x11 + ix].encode('hex'))

        ix = 0x40
        output_file.write("nand vendor:");
        while int(content[ix].encode('hex'), 16) != 0:
            output_file.write(content[ix])
            ix = ix + 1
        output_file.write("\r\n")

        ix = 0x50
        output_file.write("nand type:");
        while int(content[ix].encode('hex'), 16) != 0:
            output_file.write(content[ix])
            ix = ix + 1
        output_file.write("\r\n\r\n")
        num = num+1
#print num

File: flashInfo/flash_info.c
/**
 * made by zsir
 */
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define BLOCK_SIZE 512

int main(int argc, char *argv[])
{
    FILE *fp_read, *fp_write;
    char read_buffer[512];
    int nand_id_number;
    int file_size;
    int ix,jx;


    fp_read = fopen("./NANDINFO.nni", "r");
    if(fp_read == NULL)
    {
        return 1;
    }

    fp_write = fopen("./nandinfo.txt", "w+");
    if(fp_write == NULL)
    {
        printf("fopen fail!\n");
        return 1;
    }

    fseek(fp_read, 0, SEEK_END);
    file_size = ftell(fp_read);
    printf("file_size:%d\r\n", file_size);
    for(jx = 0; (jx + 1) * BLOCK_SIZE <= file_size; jx++)
    {
        fseek(fp_read, jx * BLOCK_SIZE, SEEK_SET);
        fread(&read_buffer, BLOCK_SIZE, 1, fp_read);
        nand_id_number = read_buffer[16];

        fprintf(fp_write, "%s", "nand id: ");

        for(ix = 0; ix < nand_id_number; ix++)
        {
            if(ix != nand_id_number - 1)
            {
                fprintf(fp_write, "0x%02x", read_buffer[17 + ix] & 0x000000ff);
                fprintf(fp_write, "%s", "-");
            }
            else
            {
                fprintf(fp_write, "%02x", read_buffer[17 + ix] & 0x000000ff);
            }
        }

        fprintf(fp_write, "%s", "\r\n");
        fprintf(fp_write, "%s", "nand厂家: ");
        fprintf(fp_write, "%s", &read_buffer[0x40]);

        fprintf(fp_write, "%s", "\r\n");
        fprintf(fp_write, "%s", "nand型号: ");
        fprintf(fp_write, "%s\r\n\r\n", &read_buffer[0x50]);
    }
    fclose(fp_read);
    fclose(fp_write);

    return 0;
}
